src/reflect/scala/reflect/internal/Variances.scala

```diff
@@ -206,13 +206,14 @@ trait Variances {
 
   /** Compute variance of type parameter `tparam` in all types `tps`. */
   def varianceInTypes(tps: List[Type])(tparam: Symbol): Variance =
-    fold(tps map (tp => varianceInType(tp)(tparam)))
+    foldMap(tps)(tp => varianceInType(tp)(tparam))
```

Comment:
I think you'd be better off with hand made tail-recursive fold here as that would avoid boxing the value class `Variance`.

```scala
def varianceInTypes(tps: List[Type])(tparam: Symbol): Variance = {
  @tailrec def loop(v: Variance, rest: List[Type]): Variance =  rest match {
      case Nil => v
      case head :: tail => loop(v & varianceInType(head), tail)
  }
}
```

Reply:
... and remove `def fold` as it would be unused after that.

Reply:
If we used a tail-recursive implementation for the `foldMap` method in the `Variance` object, which is generic on the source type, would that also perform allocations of the Variance class? 

```Scala
  def foldMap[A](as: List[A])(varianceOf: A => Variance): Variance = {
  @tailrec def loop(v: Variance, rest: List[A]): Variance = rest match {
     case Nil => v
     case h :: t => loop(v & varianceOf(h), t)
  }
}
```

Reply:
In doubt, I have re-implemented each of the four fold-using functions using a tail-recursive first-order monomorphic loop. This should avoid allocations?

Reply:
Unfortunately, if a value class is used as a function input/output type, it needs to be boxed.

This even seems to be the case with the optimizer enabled in the backend:

```
$ scala-ref 2.13.x -opt:l:inline -opt-inline-from:**/* -opt-warnings
Welcome to Scala 2.13.0-20190309-014852-2c32023 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_172).
Type in expressions for evaluation. Or try :help.

scala> class V(val x: Int) extends AnyVal; final class VClient { @inline def foo[A](f: V => A) = f(new V(0)); def test = foo(_.x.toString) }
defined class V
defined class VClient

scala> :javap -c VClient
Compiled from "<console>"
public final class $line3.$read$$iw$$iw$VClient {
  public <A> A foo(scala.Function1<$line3.$read$$iw$$iw$V, A>);
    Code:
       0: aload_1
       1: new           #14                 // class $line3/$read$$iw$$iw$V
       4: dup
       5: iconst_0
       6: invokespecial #24                 // Method $line3/$read$$iw$$iw$V."<init>":(I)V
       9: invokeinterface #30,  2           // InterfaceMethod scala/Function1.apply:(Ljava/lang/Object;)Ljava/lang/Object;
      14: areturn

  public java.lang.String test();
    Code:
       0: new           #14                 // class $line3/$read$$iw$$iw$V
       3: dup
       4: iconst_0
       5: invokespecial #24                 // Method $line3/$read$$iw$$iw$V."<init>":(I)V
       8: invokestatic  #39                 // Method $anonfun$test$1$adapted:(Ljava/lang/Object;)Ljava/lang/String;
      11: areturn

  public static final java.lang.String $anonfun$test$1(int);
    Code:
       0: iload_0
       1: invokestatic  #48                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
       4: invokevirtual #51                 // Method java/lang/Object.toString:()Ljava/lang/String;
       7: areturn

  public $line3.$read$$iw$$iw$VClient();
    Code:
       0: aload_0
       1: invokespecial #55                 // Method java/lang/Object."<init>":()V
       4: return

  public static final java.lang.String $anonfun$test$1$adapted(java.lang.Object);
    Code:
       0: aload_0
       1: checkcast     #14                 // class $line3/$read$$iw$$iw$V
       4: invokevirtual #59                 // Method $line3/$read$$iw$$iw$V.x:()I
       7: invokestatic  #64                 // Method java/lang/Integer.valueOf:(I)Ljava/lang/Integer;
      10: invokevirtual #51                 // Method java/lang/Object.toString:()Ljava/lang/String;
      13: areturn
}

scala>
```